### PR TITLE
Update builder marketplace API

### PR DESCRIPTION
**Pull request by ss-es**
### This PR: 
Changes the builder marketplace API to include a parent view and parent commitment.

This made the fallback logic for returning an empty block even more unwieldy, so this PR also includes some changes to the transactions task marketplace logic.

### CODE REVIEW

**Review by elliedavidson — COMMENTED** [review]
This looks good to me

File: crates/task-impls/src/transactions.rs [hunk 1]
```diff
@@ -232,114 +232,102 @@ impl<TYPES: NodeType, I: NodeImplementation<TYPES>, V: Versions> TransactionTask
...
 
-    #[allow(clippy::too_many_lines)]
-    /// marketplace view change handler
-    pub async fn handle_view_change_marketplace(
+    /// Produce a block by fetching auction results from the solver and bundles from builders.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if the solver cannot be contacted, or if none of the builders respond.
+    pub async fn produce_block_marketplace(
```
Comment (QuentinI) [inline crates/task-impls/src/transactions.rs:240]: Nit: doesn't need to be pub

File: crates/task-impls/src/transactions.rs [hunk 2]
```diff
@@ -232,114 +232,102 @@ impl<TYPES: NodeType, I: NodeImplementation<TYPES>, V: Versions> TransactionTask
...
-                for bundle in join_all(futures).await {
-                    match bundle {
-                        Ok(Ok(b)) => bundles.push(b),
-                        _ => continue,
-                    }
-                }
+        for bundle in join_all(futures).await {
+            match bundle {
+                Ok(Ok(b)) => bundles.push(b),
+                _ => continue,
```
Comment (QuentinI) [inline crates/task-impls/src/transactions.rs:291]: Out of scope for this, so feel free to ignore, but it'd be nice to have some logging of the timeout or error here.

File: crates/task-impls/src/transactions.rs [hunk 3]
```diff
@@ -232,114 +232,102 @@ impl<TYPES: NodeType, I: NodeImplementation<TYPES>, V: Versions> TransactionTask
...
+        ))
+    }
 
+    /// Produce a null block
+    pub async fn null_block(
+        &mut self,
+        block_view: TYPES::Time,
+        version: Version,
+    ) -> Option<PackedBundle<TYPES>> {
         // Increment the metric for number of empty blocks proposed
```
Comment (QuentinI) [inline crates/task-impls/src/transactions.rs:331]: Nit: I'd make this function sync and move the metric increment to the caller to make this one side-effect free

File: crates/task-impls/src/transactions.rs [hunk 4]
```diff
@@ -361,16 +349,51 @@ impl<TYPES: NodeType, I: NodeImplementation<TYPES>, V: Versions> TransactionTask
...
+    ) -> Option<HotShotTaskCompleted> {
+        let version = match self.upgrade_lock.version(block_view).await {
+            Ok(v) => v,
+            Err(err) => {
+                error!("Upgrade certificate requires unsupported version, refusing to request blocks: {}", err);
+                return None;
+            }
+        };
+
+        let (parent_view, parent_hash) = self.last_vid_commitment(block_view).await?;
```
Comment (QuentinI) [inline crates/task-impls/src/transactions.rs:377]: Can we add a retry loop on this similar to what we have in legacy? It's not uncommon for the leader not to have the relevant VID right away.
Also, if we fail to find it, we should still produce an empty block.

Comment (ss-es) [reply crates/task-impls/src/transactions.rs]: Updated: I created a wrapper `last_vid_commitment_retry` for the loop, and used that in both places.

**Review by QuentinI — APPROVED** [review]
LGTM!

**Pull request merged**